[bug] Premake generator is broken

### Describe the bug

Conan version 2.9.2


### How to reproduce it

when calling the generated `conan_setup()` in `conandeps.premake5.lua` without arguments,
it only works with x86 release build because in `conan_setup_build()` it calls
`filter { "configurations:release", "architecture:x86_64" }`
which filters away all the setup calls for any other configuration.

when i call `conan create .` on my recipe it seems to default to debug build,
because it fails to add dependencies' include dirs.

it's possible to work around the wrong config detection by manually specifying
the configuration string (example: `conan_setup("release_x86_64")`,
for every configuration combination...

[bug_premake.zip](https://github.com/user-attachments/files/17819834/bug_premake.zip)

Conan 1's Premake generator worked fine, if it ain't broke why break it?

when I run `conan create .` the package build silently(?) fails.
take note of the `Building nanovg (debug)`, which mismatches the Conan generator's release (the generated Premake Lua scripts only contain release config)?
```
nanovg/master: Calling build()
nanovg/master: RUN: premake5 gmake2
Building configurations...
Running action 'gmake2'...
Generated Makefile...
Generated nanovg.make...
Done (18ms).

nanovg/master: RUN: build
==== Building nanovg (debug) ====
Creating obj/Debug
Creating bin/Debug
nanovg.c
In file included from nanovg/src/nanovg.c:26:
nanovg/src/fontstash.h:150:10: fatal error: ft2build.h: No such file or directory
  150 | #include <ft2build.h>
      |          ^~~~~~~~~~~~
compilation terminated.
make[1]: *** [nanovg.make:142: obj/Debug/nanovg.o] Error 1
make: *** [Makefile:30: nanovg] Error 2
```

it seems to be silent because it carries on to run the test_package which also fails because there's no compiled library:
```
======== Testing the package: Building ========
nanovg/master (test package): Calling build()
nanovg/master (test package): Running CMake.configure()
nanovg/master (test package): RUN: cmake -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE="generators/conan_toolchain.cmake" -DCMAKE_INSTALL_PREFIX="/mnt/hdd/dev_hdd/nanovg/conan-nanovg/test_package" -DCMAKE_POLICY_DEFAULT_CMP0091="NEW" -DCMAKE_BUILD_TYPE="Release" "/mnt/hdd/dev_hdd/nanovg/conan-nanovg/test_package"
-- Using Conan toolchain: /mnt/hdd/dev_hdd/nanovg/conan-nanovg/test_package/build/gcc-14-x86_64-17-release/generators/conan_toolchain.cmake
-- Conan toolchain: Defining architecture flag: -m64
-- Conan toolchain: C++ Standard 17 with extensions OFF
-- The CXX compiler identification is GNU 14.2.1
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Conan: Target declared 'nanovg::nanovg'
CMake Error at build/gcc-14-x86_64-17-release/generators/cmakedeps_macros.cmake:67 (message):
  Library 'nanovg' not found in package.  If 'nanovg' is a system library,
  declare it with 'cpp_info.system_libs' property
Call Stack (most recent call first):
  build/gcc-14-x86_64-17-release/generators/nanovg-Target-release.cmake:23 (conan_package_library_targets)
  build/gcc-14-x86_64-17-release/generators/nanovgTargets.cmake:24 (include)
  build/gcc-14-x86_64-17-release/generators/nanovg-config.cmake:16 (include)
  CMakeLists.txt:5 (find_package)


-- Configuring incomplete, errors occurred!
```

this happens because in the generated premake file there's only support for 1 configuration which mismatches what's actually being used.
for example in `conan_freetype_vars_release_x86_64.premake5.lua`:
```lua
-- ...
t_conandeps["release_x86_64"]["freetype"] = {}
t_conandeps["release_x86_64"]["freetype"]["includedirs"] = {"/home/desktop/.conan2/p/b/freete219fd5cdef6f/p/include",
"/home/desktop/.conan2/p/b/freete219fd5cdef6f/p/include/freetype2"}
t_conandeps["release_x86_64"]["freetype"]["libdirs"] = {"/home/desktop/.conan2/p/b/freete219fd5cdef6f/p/lib"}
-- ...
```

and in `conandeps.premake5.lua`:
```lua
function conan_setup_build(conf, pkg)
    if conf == nil then
        filter { "configurations:release", "architecture:x86_64" } -- defaults to only working with this configuration
            conan_setup_build("release_x86_64")
        filter {}
    elseif pkg == nil then
        for k,v in pairs(conandeps[conf]) do
            conan_setup_build(conf, k)
        end
    else

        includedirs(conandeps[conf][pkg]["includedirs"])
        bindirs(conandeps[conf][pkg]["bindirs"])
        defines(conandeps[conf][pkg]["defines"])

    end
end
```
Why the generated Premake scripts are release/debug and architecture specific?
with the Conan 1 generator they weren't and you'd use separate build directory for each configuration,
which avoids the need to manually select the 1 generated config in the `premake5.lua` file.

i don't understand what this new generator tries to solves, it only seems to over-complicate things with all these nested dictionaries, merging and recursion.
the old generator just generate a list of includedirs/defines/etc and a function that adds them to Premake.

--- a/conan/tools/premake/premake.py
+++ b/conan/tools/premake/premake.py
@@ -1,39 +1,17 @@
-class Premake(object):
+class Premake:
     def __init__(self, conanfile):
         self._conanfile = conanfile
 
     # automatically chooses premake action based on used compiler
     def configure(self):
-        if "Visual Studio" in self._conanfile.settings.compiler:
-            _visuals = {'8': '2005',
-                        '9': '2008',
-                        '10': '2010',
-                        '11': '2012',
-                        '12': '2013',
-                        '14': '2015',
-                        '15': '2017',
-                        '16': '2019'}
-            premake_command = "premake5 vs%s" % _visuals.get(str(self._conanfile.settings.compiler.version))
-            self._conanfile.run(premake_command)
-        elif "msvc" in self._conanfile.settings.compiler:
-            _visuals = {'14.0': '2005',
-                        '15.0': '2008',
-                        '16.0': '2010',
-                        '17.0': '2012',
-                        '18.0': '2013',
-                        '19.0': '2015',
-                        # add non-trailing 0 variation manually
-                        '19.1': '2017',
-                        '19.2': '2019'}
-            # generate VS2017 versions
-            for i in range(0,7):
-                ver = '19.1' + str(i)
-                _visuals[ver] = '2017'
-            # generate VS2019 versions
-            for i in range(0,10):
-                ver = '19.2' + str(i)
-                _visuals[ver] = '2019'
-            premake_command = "premake5 vs%s" % _visuals.get(str(self._conanfile.settings.compiler.version))
+        if "msvc" in self._conanfile.settings.compiler:
+            _visuals = {'190': '2015',
+                        '191': '2017',
+                        '192': '2019',
+                        '193': '2022',
+                        '194': '2023'}
+            version = _visuals.get(str(self._conanfile.settings.compiler.version))
+            premake_command = f"premake5 vs{version}"
             self._conanfile.run(premake_command)
         else:
             self._conanfile.run("premake5 gmake2")

--- a/conan/tools/premake/premakedeps.py
+++ b/conan/tools/premake/premakedeps.py
@@ -110,7 +110,7 @@ class _PremakeTemplate(object):
             if dep_cpp_info.sysroot else ""
 
 
-class PremakeDeps(object):
+class PremakeDeps:
     """
     PremakeDeps class generator
     conandeps.premake5.lua: unconditional import of all *direct* dependencies only
@@ -192,9 +192,7 @@ class PremakeDeps(object):
         build_req = self._conanfile.dependencies.build
 
         # Merge into one list
-        full_req = list(host_req.items()) \
-                   + list(test_req.items()) \
-                   + list(build_req.items())
+        full_req = list(host_req.items()) + list(test_req.items()) + list(build_req.items())
 
         # Process dependencies and accumulate globally required data
         pkg_files = []

--- a/test/conftest.py
+++ b/test/conftest.py
@@ -164,6 +164,7 @@ tools_locations = {
                            'Darwin': '/Users/jenkins/bazel-7.1.2/bin'}},
     },
     'premake': {
+        "disabled": True,
         "exe": "premake5",
         "default": "5.0.0",
         "5.0.0": {

--- a/test/functional/toolchains/test_premake.py
+++ b/test/functional/toolchains/test_premake.py
@@ -1,0 +1,74 @@
+import platform
+import textwrap
+
+import pytest
+
+from conan.test.assets.sources import gen_function_cpp
+
+
+@pytest.mark.skipif(platform.system() == "Darwin", reason="Not for MacOS")
+@pytest.mark.tool("premake")
+def test_premake(matrix_client):
+    c = matrix_client
+    conanfile = textwrap.dedent("""
+        import os
+        from conan import ConanFile
+        from conan.tools.premake import Premake
+        from conan.tools.microsoft import MSBuild
+        class Pkg(ConanFile):
+            settings = "os", "compiler", "build_type", "arch"
+            requires = "matrix/1.0"
+            generators = "PremakeDeps", "VCVars"
+            def build(self):
+                p = Premake(self)
+                p.configure()
+                build_type = str(self.settings.build_type)
+                if self.settings.os == "Windows":
+                    msbuild = MSBuild(self)
+                    msbuild.build("HelloWorld.sln")
+                else:
+                    self.run(f"make config={build_type.lower()}_x86_64")
+                p = os.path.join(self.build_folder, "bin", build_type, "HelloWorld")
+                self.run(f'"{p}"')
+        """)
+    premake = textwrap.dedent("""
+        -- premake5.lua
+
+        include('conandeps.premake5.lua')
+
+        workspace "HelloWorld"
+           conan_setup()
+           configurations { "Debug", "Release" }
+           platforms { "x86_64" }
+
+        project "HelloWorld"
+           kind "ConsoleApp"
+           language "C++"
+           targetdir "bin/%{cfg.buildcfg}"
+
+           files { "**.h", "**.cpp" }
+
+           filter "configurations:Debug"
+              defines { "DEBUG" }
+              symbols "On"
+
+           filter "configurations:Release"
+              defines { "NDEBUG" }
+              optimize "On"
+
+           filter "platforms:x86_64"
+              architecture "x86_64"
+          """)
+    c.save({"conanfile.py": conanfile,
+            "premake5.lua": premake,
+            "main.cpp": gen_function_cpp(name="main", includes=["matrix"], calls=["matrix"])})
+    c.run("build .")
+    assert "main: Release!" in c.out
+    assert "matrix/1.0: Hello World Release!" in c.out
+    if platform.system() == "Windows":
+        assert "main _M_X64 defined" in c.out
+    else:
+        assert "main __x86_64__ defined" in c.out
+    c.run("build . -s build_type=Debug --build=missing")
+    assert "main: Debug!" in c.out
+    assert "matrix/1.0: Hello World Debug!" in c.out

